### ISSUE TITLE
Question about `maskrcnn_benchmark/modeling/roi_heads/relation_head/loss.py`

### ISSUE THREAD

**Issue by MengHsir** [guikunchen/SDSGG#5]
（1）关于`self.description_relation_loss`在`__init__`中正确索引问题，以及关于`target[:, 21] = target[:, 21] * 0.5`的疑惑。
代码当中，会根据使用的base谓词索引从读入的description_relation_loss.csv数据当中对应的行提取数据，保留的`self.description_relation_loss`数据只与`TRAIN_PART==“base”`有关。但是在forward函数当中`target = torch.cat([(self.description_relation_loss[x, 2]).unsqueeze(0) for x in target]).half()`存在如下问题：
传入的target=rel_labels的索引是self.base=[0, 1, 3, 6, 7, 8, 9, 10, 11, 12, 13, 16, 17, 19, 20, 21, 22, 23, 25, 27, 29, 30, 31, 33, 35, 37, 38, 40, 41, 42, 43, 46, 47, 48, 49, 50]，此时，如果x=3，那么self.description_relation_loss[x, 2]=self.description_relation_loss[3, 2],此时self.description_relation_loss第三行对应的并不是原始description_relation_loss.csv当中索引为3：'against': 3的谓词loss内容,是否是因为在visual_genome.py当中执行了`_filter_labels`操作。

（2）并且，我不明白为什么`self.description_relation_loss[x, 2]`要选择2，且为什么第21个描述数据要缩小一半：`target[:, 21] = target[:, 21] * 0.5`，这些内容我在论文当中并没有看到，希望作者可以解疑答惑🙏🙏🙏🙏🙏🙏

（3）最后，对于预测器当中的`probs=(similarity1-similarity2)/0.05`参数0.05的选择是否有某种意义，还是单纯遵循了CLIP的温度缩放系数？

```
class Loss(nn.Module):
    def __init__(self, gamma=0, alpha=None, size_average=True, device=None):
        super(Loss, self).__init__()
        self.gamma = gamma
        global PRDCS_BASE,PRDCS_NOVEL,SEMAN,datamode

        self.alpha = alpha
        self.size_average = size_average

        self.description_relation_loss = pd.read_csv(curpath+"/description_relation_loss.csv")
        self.id_dict = {'__background__': 0, 'above': 1, 'across': 2, 'against': 3, 'along': 4, 'and': 5, 'at': 6,
                        'attached to': 7, 'behind': 8, 'belonging to': 9, 'between': 10, 'carrying': 11,
                        'covered in': 12, 'covering': 13, 'eating': 14, 'flying in': 15, 'for': 16, 'from': 17,
                        'growing on': 18, 'hanging from': 19, 'has': 20, 'holding': 21, 'in': 22, 'in front of': 23,
                        'laying on': 24, 'looking at': 25, 'lying on': 26, 'made of': 27, 'mounted on': 28, 'near': 29,
                        'of': 30, 'on': 31, 'on back of': 32, 'over': 33, 'painted on': 34, 'parked on': 35,
                        'part of': 36, 'playing': 37, 'riding': 38, 'says': 39, 'sitting on': 40, 'standing on': 41,
                        'to': 42, 'under': 43, 'using': 44, 'walking in': 45, 'walking on': 46, 'watching': 47,
                        'wearing': 48, 'wears': 49, 'with': 50}

        if DATA_OPTION == "vg":
            self.base=[0,1, 3, 6, 7, 8, 9, 10, 11, 12, 13, 16, 17, 19, 20, 21, 22, 23, 25, 27, 29, 30, 31, 33, 35, 37, 38, 40, 41, 42, 43, 46, 47, 48, 49, 50]
        elif DATA_OPTION == "gqa":
            self.base = [0, 31, 48, 30, 29, 22, 8, 23, 21, 1, 50, 40, 43, 38, 41, 11, 46, 6, 13, 35, 47, 12]  # gqa

        if TRAIN_PART=="base":
            self.description_relation_loss = self.description_relation_loss.iloc[self.base, 1:]
        elif TRAIN_PART=="total":
            self.description_relation_loss = self.description_relation_loss.iloc[:, 1:]


        self.description_relation_loss = self.description_relation_loss.applymap(lambda x: [int(s) for s in x.split(',')])
        self.description_relation_loss = np.array(self.description_relation_loss)
        self.description_relation_loss = np.array([[np.array(item) for item in inner_list] for inner_list in self.description_relation_loss])
        self.description_relation_loss = torch.Tensor(self.description_relation_loss).to(device)

    def forward(self, input, target):

        target = target.view(-1)

        zz = torch.where(target > 0, True, False)
        input = input[zz].half()
        target = target[zz]

        totarget=input[:,1]
        input=input[:,0]

        target = torch.cat([(self.description_relation_loss[x, 2]).unsqueeze(0) for x in target]).half()
        target[:, 21] = target[:, 21] * 0.5

        target=target*2+totarget*0.01-0.03

        loss = F.mse_loss(input, target, reduction="mean", reduce=True).half()
        return loss
```

**Comment by MengHsir** [comment]
希望作者可以回复一下这个关于loss计算的关键问题，对我很重要，我会非常感谢作者的回复！ @guikunchen

**Comment by ln172** [comment]
不好意思，最近很忙。
（1）对于1这个问题，_filter_labels 重新映射了索引，在代码maskrcnn_benchmark/data/datasets/visual_genome.py部分。
（2）关于为什么self.description_relation_loss[x, 2]要选择2，在description_relation_loss.csv这个文件中，我们没有进行过滤，也就是说每一列其实是相同的，选取第一列之后的任意列即可。对于为什么第21个描述数据要缩小一半，这个描述并非视觉描述，而是用来区分主谓语的，只是一个标记，此点在论文中可以看见。
（3）只是遵循了CLIP的温度缩放系数。

**Comment by MengHsir** [comment]
感谢作者！解答了我很久的困惑！！！此外，代码当中有一处与论文当中不同的错误：
`target=target*2+totarget*0.01-0.03`，代码当中超参数beta=0.1,代码当中可能不小心定义为0.01

<img width="967" height="395" alt="Image" src="https://github.com/user-attachments/assets/1ae7600d-1531-48e0-bfb1-829cd6e8a012" />
